design-proposals/klip-19-materialize-views.md

```diff
@@ -0,0 +1,223 @@
+ 
+## What is not in scope
+
+* No `VIEW` functionality, only `MATERIALIZED VIEW`s. `VIEW`s have the potential of being useful, as they would allow
+  user to define multiple statements that can later be materialized in a single KS topology. But such work is outside the 
+  scope of this KLIP.
+* New `CREATE TABLE AS SELECT` functionality. This can be done later.
+
+## Value/Return
+
+Differentiating between mutable and read-only data sources will make KSQL easier for both engineers and end-users
+to reason about, the SQL more explicit and is needed to to fix the inconsistencies in `INSERT VALUES` functionality.  
+  
+## Public APIS
+
+We propose the following SQL syntax and associated functionality:
+
+### Tables:
+
+* Add `CREATE MATERIALIZED VIEW` to import an existing topic containing a table. 
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` *not* supported in `WITH` clause.
+* Re-purpose `CREATE TABLE` to create a mutable table backed by a new or existing topic.
+    * `INSERT`s into the table will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE TABLE AS SELECT` with `CREATE MATERIALIZED VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Remove `CREATE TABLE AS SELECT`.
+    
+For those worried about the excessive typing required for `MATERIALIZED VIEW`, we could add a shorthand `CREATE MV` etc. 
+
+### Streams:
+
+Standard SQL already has `MATERIALIZED VIEW`, but has no stream equivalent. 
+We propose a stream equivalent of `MATERIALIZED STREAM VIEW`.
+
+* Add `CREATE MATERIALIZED STREAM VIEW` to import an existing topic containing a stream.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` *not* supported in `WITH` clause.
+* Re-purpose `CREATE STREAM` to create a mutable stream backed by a new or existing topic.
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of any new topic. 
+* Replace `CREATE STREAM AS SELECT` with `CREATE MATERIALIZED STREAM VIEW AS SELECT`.
+    * `INSERT`s will *not* be allowed into the view, as views are read-only.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.   
+* Either `CREATE STREAM AS SELECT` will be removed, or maintained as creating a mutable derived stream. 
+    * `INSERT`s into the stream will be supported.
+    * `PARTITIONS` and `REPLICAS` supported in `WITH` clause to control customisation of the sink topic.
+    
+For those worried about the excessive typing required for `MATERIALIZED STREAM VIEW`, we could add a shorthand `CREATE MSV` etc.
```

Comment:
Not sure about this...

Reply:
It's only an option/possibility.  Could be implemented via CLI's auto complete for example.

Reply:
Auto-complete seems a good option -- but I would not make it part of the language.

Reply:
These kinds of things won't be typed very often so optimising for the time taken here doesn't make a lot of sense imo.

Reply:
removed!